text/3617-precise-capturing.md

```diff
@@ -0,0 +1,1063 @@
+[guide-level-explanation]: #guide-level-explanation
+
+In all editions, RPIT-like `impl Trait` opaque types may include `use<..>` before any bounds to specify which in-scope generic parameters are captured or that no in-scope generic parameters are captured (with `use<>`).  If `use<..>` is provided, it entirely overrides the implicit rules for which generic parameters are captured.
+
+One way to think about `use<..>` is that, in Rust, `use` brings things *into scope*, and here we are bringing certain generic parameters into scope for the hidden type.
+
+For example, we can solve the overcapturing in the original motivating example by writing:
+
+```rust
+fn foo<T>(_: T) -> impl use<> Sized {}
+//                 ^^^^^^^^^^^^^^^^
+//                 ^ Captures nothing.
+```
+
+Similarly, we can use this to avoid overcapturing a lifetime parameter so as to migrate code to Rust 2024:;
+
+```rust
+fn foo<'t>(_: &'t ()) -> impl use<> Sized {}
+//                       ^^^^^^^^^^^^^^^^
+//                       ^ Captures nothing.
+```
+
+We can use this to capture some generic parameters but not others:
+
+```rust
+fn foo<'t, T, U>(_: &'t (), _: T, y: U) -> impl use<U> Sized { y }
+//                                         ^^^^^^^^^^^^^^^^^
+//                                         ^ Captures `U` only.
+```
+
+## Generic const parameters
+
+In addition to type and lifetime parameters, we can use this to capture generic const parameters:
+
+```rust
+fn foo<'t, const C: u8>(_: &'t ()) -> impl use<C> Sized { C }
+//                                    ^^^^^^^^^^^^^^^^^
+//                                    ^ Captures `C` only.
+```
+
+## Capturing from outer inherent impl
+
+We can capture generic parameters from an outer inherent impl:
+
+```rust
+struct Ty<'a, 'b>(&'a (), &'b ());
+
+impl<'a, 'b> Ty<'a, 'b> {
+    fn foo(x: &'a (), _: &'b ()) -> impl use<'a> Sized { x }
+    //                              ^^^^^^^^^^^^^^^^^^
+    //                              ^ Captures `'a` only.
+}
+```
+
+## Capturing from outer trait impl
+
+We can capture generic parameters from an outer trait impl:
+
+```rust
+trait Trait<'a, 'b> {
+    type Foo;
+    fn foo(_: &'a (), _: &'b ()) -> Self::Foo;
+}
+
+impl<'a, 'b> Trait<'a, 'b> for () {
+    type Foo = impl use<'a> Sized;
+    //         ^^^^^^^^^^^^^^^^^^
+    //         ^ Captures `'a` only.
+    fn foo(x: &'a (), _: &'b ()) -> Self::Foo { x }
+}
+```
+
+## Capturing in trait definition
+
+We can capture generic parameters from the trait definition:
+
+```rust
+trait Trait<'a, 'b> {
+    fn foo(_: &'a (), _: &'b ()) -> impl use<'a, Self> Sized;
+    //                              ^^^^^^^^^^^^^^^^^^^^^^^^
+    //                              ^ Captures `'a` and `Self` only.
+}
+```
+
+## Capturing elided lifetimes
+
+We can capture elided lifetimes:
+
+```rust
+fn foo(x: &()) -> impl use<'_> Sized { x }
+//                ^^^^^^^^^^^^^^^^^^
+//                ^ Captures `'_` only.
+```
```

Comment:
This is not clearly documented about what "captures `'_` only" means. It reads like it captures all elided lifetimes, but that's not the case.

After [some testing on playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fd990087c19093c8e216a896104c9095), it turns out `'_` works as the same way as the `'_` in a normal result type, that is:
- Refer to the only elided lifetime, if there is only a single elided lifetime and no named lifetimes.
  - So you cannot use `'_` where there are named lifetimes, even if only a single lifetime is elided.
- Refer to the elided lifetime of `&self` if there is one. Other elided lifetimes are not captured.
- Otherwise, report an ambiguity error.

Ideally we should document this somewhere.